README
ABOUT LISPEL

 This is Lispel, a (very) small interpreter for a Lisp-like Extension Language.
 It is aimed at providing plugin scripting support for C++ applications.

 I wrote this interpreter because I wanted to have scripting capabilities
 for my C++ software written for a wide range of platforms. The fact that
 this software was among others to support DOS ruled out a number of
 candidates (normally I would have used Tcl/Tk to add scripting support).
 So I decided to write a small interpreter from scratch, implementing a 
 Lisp-like language (more exactly, a Scheme-like language, as what today is
 known as Lisp is much too bloated).

 Although the Lispel distribution provides a small interactive interpreter
 Lispel is in no way suitable for writing large scale applications solely on
 the scripting level. It is, among other deficiencies, much too slow for such
 a task. On the other hand, if you wanted to build a application based on some
 library, you could start by providing scripting support for that library and
 build a prototype using scripts.


VERSION AND STATUS

 The current version, Lispel 0.0.3, really should be regarded as some kind of
 preview. Many things that need to be implemented are still missing and some
 things found in 'real' Scheme interpreters will never be implemented.
 Among the missing features from the first category mentioned are:
 yet. These include:

 - You can't use the quoting character, you will have to use the quote special
   form explicitly, instead (a bug!).
 - The scanner doesn't recognize numeric constants with a radix other than 10
   and character constants.
 - Dotted pairs aren't supported (this might belong to the second category).
 - Continuations aren't implemented, they'll need a new evaluator
   implementation.

 During the next releases the various components making up the interpreter
 will be gradually enhanced or replaced by improved implementations. For
 forthcoming additions, refer to the file TODO. But before I will integrate
 a witty super-compiler, full CORBA scripting support and a nuclear power plant
 control module I will keep on adding basic data types and builtins. Then the
 components will be improved bottom up, which basically means that I will start
 improvements from the scanner class up to the class Interpreter.


FEATURES

 Yes, there are also some features. Some of them are:

 - Writing builtins is easier than in most other embeddable scripting 
   languages. Normally, when implementing a builtin for some scripting language
   you will first check the number of arguments, then you have to check if each
   of the arguments has the correct type and then you can write down the code
   for what you really wanted to do. This approach adds much overhead to each
   command implementation and can be a source of nasty bugs during development.
   In Lispel, the type checking is done automatically behind the scenes.
 - Hash tables aren't required by the Scheme standard, but as I regard them as
   an important data structure especially for scripting languages they are
   supported by Lispel.
 - The code is very simple so that you can have a jump start if you want to
   add or modify anything. Or you can simply 'steal' some of the components
   making up Lispel to implement your own scripting language.


PLATFORMS

 Lispel has been successfully compiled on DOS, Windows and Linux using
 egcs 1.1.2 and gcc 2.95.2. Egcs 1.1.2 likes to issue strange warnings about
 uninitialized variables because of it's poor code flow analysis in presence
 of exceptions. This problem will probably be worse when using older compiler
 versions (which I haven't tested). Aside from this, Lispel should be fairly
 portable as long as you use a compiler that fully supports STL as the sources
 are written in pure ANSI-C++ (at least I think so).


FEATURES / DESIGN

 Not that many features and an evolving design. ;-)
 

FUTURE DIRECTIONS

 The items described in the following paragraphs won't be added before the
 Lispel implementation is stable and complete. 'Stable' and 'complete' are
 very vague terms which I used intentionally because I am currently unsure
 about how far Lispel should deviate from mainstream Scheme implementations
 to be able to fully support new features.

 The first extension package for Lispel will be an XML processing package
 because this is a part of the project I wrote Lispel for. Other parts of this
 project will include constraint propagation and neural network computations.
 The XML processing facility will be based on the GPL'ed Gnome XML library.

 One thing that should be relatively easy to accomplish is the
 implementation of new syntaxes for Lispel by essetially replacing the reader
 class. This would allow a user to select a preferred syntax, may it be 
 lisp-like or Javascript, during interpreter initialization.


CODING

 If you would like to contribute some code then please try to stick to the
 coding conventions explicated below. They lack some details, but you should
 get an impression of what's going on by skimming over the sources (what you
 will probably do anyway when you start extending the interpreter).

 * Avoid compiler warnings: the compiler doesn't issue warning because it is
   afraid of getting bored, so don't ignore them!

 * Use the new typecast constructs, they make your intentions much clearer and
   thus yield code which more readable for others

 * It is always a good idea to let Emacs indent your code (I personally like
   my code indented --braces-on-if-line, but don't feel forced into sticking
   to this style)

 * If you think that there are good reasons to break the coding conventions
   described above then feel free to do so.
 
 * Prefix member variables with m_ to avoid clashes with parameter and method
   names. For static member variables the prefix M_ should be used to make
   clearer that they are another type of member variables.

 * Use dynamic binding and factory classes for components that you expect to
   be extended or that may be replaced by alternative components. If you
   build more complex subsystems, provide a common (mid-level) API by 
   implementing a facade class.
 
 * I try to avoid implicit conversions wherever possible as this makes the 
   code much easier to understand and hence much safer. Use the explicit 
   keyword where appropriate and be careful with defining conversion operators.


REFERENCES

 There are some web sites dedicated to Scheme which keep many pointers to 
 introductory material, other Scheme implementations, ...
 Most of the sites can be found in the Scheme FAQ, available at
  ftp.cs.cmu.edu:/user/ai/pubs/faqs/scheme/
 Some sites containing good introductory papers and some sample programs are
  - www.schemers.org
  - ftp.cs.indiana.edu:/pub/scheme-repository/
  - ftp.cs.cmu.edu:/user/ai/lang/scheme/
  - swiss-ftp.ai.mit.edu:/archive/scheme-reports/
  - faui80.informatik.uni-erlangen.de/pub/scheme/

 A good reference of the Scheme language is the 'Revised 4th Report on Scheme',
 often referred to as R4RS, which many features of Lispel are based on. The
 document LANGUAGE describes the major differences between Scheme and Lispel.
 Additionally, the sample programs in the directory samples/lispel should
 give you a first impression of the language's features. The document can be
 found at
  ftp://swiss-ftp.ai.mit.edu/pub/scm/HTML/r4rs_toc.html or
  http://www.maths.tcd.ie/scrg/scheme/index.html

 If you want to hack the interpreter then you best refer to the doc++ generated
 API documentation. If you have a copy of doc++, then simply go to the doc/api
 directory and type 'make'. The generated documentation should be available
 as a separate package from where you obtained the Lispel sources.
 Additionally, the directory samples/extension contains some simple extensions
 to the interpreter which can be used as a starting point for your own work.


CONTACT

 If you have any comments or suggestions then feel free to send a mail to
   msaure@rz.uni-osnabrueck.de
 The Lispel homepage can be found at 
   http://cmp.cl-ki.uni-osnabrueck.de/~michael/lispel/
 This might change in the near future as I don't know if I am allowed to 
 keep this page.


Have fun!



NEWS
CHANGES BETWEEN LISPEL VERSIONS
-------------------------------


* Changes from 0.0.4 to 0.0.3					unreleased

 - many, many code changes to add full ANSI-C++ compliance
 - use external MiscClass class library

* Changes from 0.0.3 to 0.0.2					unreleased

 - improved scanner with support for character constants and proper
   backslash substitution
 - the sample command line driver can now be used as a UN*X script interpreter
   (new comment introducer #!)
 - improved error reporting
 - proper handling of quoting character
 - many bugfixes
 ! this version wasn't released because of some bugs that would be fixed by
   a new implementation of some components, which is scheduled for 0.1.0


* Changes from 0.0.2 to 0.0.1					21-nov-1999

 - better exception handling: uses a tree of exception classes, uses throw
   declarations
 - hashtable data type added
 - added keywords known from ANSI Common Lisp
 - improved lambda implementation (body can be a list of expressions)
 - more than 50 builtins, important special forms cond, let, let* added


* Version 0.0.1							15-nov-1999

 - This is a very first implementation of the core of an interpreter for a
   Lisp- (or Scheme-)like extension language. It serves as a prototype, only,
   so many components will be replaced before the next major version comes out.

$Id: NEWS,v 1.1.1.1 2001/10/31 15:44:42 michael Exp $



lispel/corecmd.cpp
/*
 * corecmd.cpp ---
 *
 *   Provides the most basic commands and special forms.
 *
 * (c) 2000-2001 by Michael Saure <MichaelSaure@web.de>
 * RCSID: $Id: corecmd.cpp,v 1.1.1.1 2001/10/31 15:44:42 michael Exp $
 */

#include <list>
#include <ctime>
#include <lispel/corecmd.hh>
#include <lispel/lispel.hh>
#include "version.h"

/*
 * SPECIAL FORMS
 */

Handle_ptr
LambdaCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  MCAssertValidInstance();
  assert( 0 != env && 0 != ctx.eval && 0 != expr);

  Handle_ptr args = expr->cdr();
  if (!isList( args->car()))
    throw TypeException( "list", __FILE__, __LINE__);
  
  // collect argument names and check if argument specification is valid
  std::list<Handle_ptr> argumentList;
  if (!eq( args->car(), ctx.NIL)) {
     std::back_insert_iterator<std::list<Handle_ptr> > iiter( argumentList);
     copyList( args->car(), iiter);
     std::list<Handle_ptr>::iterator pos;
     for (pos=argumentList.begin(); pos!=argumentList.end(); ++pos) {
       if (!(*pos)->hasType( Handle::ntSYMBOL))
         throw TypeException( "symbol", __FILE__, __LINE__);
     }
  } else {
     // empty argument list
     std::cerr << "lambda: empty argument list" << std::endl;
  }

#if defined( DEBUG) && DEBUG > 3
  std::cerr << "create closure with " << argumentList.size() << " arguments."
       << std::endl;
  std::cerr << "  body: ";
  printList( args->cdr(), std::cerr); std::cerr << std::endl;
  std::cerr << "  args: ";
  std::list<Handle_ptr>::iterator pos;
  for (pos=argumentList.begin(); pos!=argumentList.end(); ++pos)
    std::cerr << *(*pos) << ' ';
  std::cerr << std::endl;
#endif   

  return ctx.factory->makeClosure( argumentList, env, args->cdr());
}


Handle_ptr
CondCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
   Handle_ptr args = expr->cdr();
   
   Handle_ptr branch;   // iterate over all clauses/branches
   for (branch=args; !eq( branch, ctx.NIL); branch=branch->cdr()) {
      Handle_ptr exprList = branch->car();
      Handle_ptr evaluated = exprList->car();

      if (!eq( ctx.eval->eval( evaluated), ctx.FALSE)) {
        // expression evaluates to true, so evaluate remaining expressions in
	// the current branch and return result of last evaluation
	for ( exprList=exprList->cdr(); !eq( exprList, ctx.NIL);
	      exprList=exprList->cdr()) 
	    evaluated = ctx.eval->eval( exprList->car());
	return evaluated;
      }
   }
   return ctx.NIL; //this is left unspecified by R[45]RS
}


Handle_ptr
QuoteCommand::execute( Context &ctx, Environment *env, Handle_ptr args)
{
  MCAssertValidInstance();
  assert( 0 != args);

  if (eq( ctx.NIL, args)) {
    throw TypeException( "non-NIL", __FILE__, __LINE__);
  }
  // MISSING: check number of arguments (list_length)
  if (eq( args->cdr(), ctx.NIL))
    throw ArgumentCountException( 1, __FILE__, __LINE__);

  return args->cdr()->car();
}
  
Handle_ptr
DefineCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  MCAssertValidInstance();

  if (3 != listlength( expr))
    throw ArgumentCountException( 3, __FILE__, __LINE__);

  Handle_ptr args = expr->cdr();
  if (args->car()->hasType( Handle::ntSYMBOL)) {
    // Variablenbindung
    ctx.toplevel->put( args->car()->stringValue(),
		       ctx.eval->eval( args->cdr()->car()));
    return args->car();

  } else
    if (args->car()->hasType( Handle::ntCONS) && !args->car()->isNilRep()) {
      // Funktionsdefinition
      std::cerr << "error: function definitions aren't supported right now" << std::endl;
      return ctx.NIL;
      
    } else
      throw TypeException( "symbol or list", __FILE__, __LINE__);
  
  MCAssertNotReached( 0);
}



Handle_ptr
LetCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  MCAssertValidInstance();
  MCAssert( 0 != expr && !eq( expr, ctx.NIL), "illegal expression");

  if (!eq( expr->cdr(), ctx.NIL) && !eq( expr->cdr()->cdr(), ctx.NIL)) {
     // create new binding environment
     Environment *newEnv = env->makeChildEnvironment();
     Handle_ptr bindList = expr->cdr()->car();

#if defined( DEBUG) && DEBUG > 2
     std::cerr << "let bindings == ";
     printList( bindList, std::cerr);
     std::cerr << std::endl;
#endif

     for ( ; !eq( bindList, ctx.NIL); bindList = bindList->cdr()) {
        Handle_ptr bindings = bindList->car();
#if defined( DEBUG) && DEBUG > 3
        std::cerr << " evaluate binding ";
	printList( bindings, std::cerr);
	std::cerr << std::endl;
#endif
	std::string key = bindings->car()->stringValue();
	Handle_ptr value = ctx.eval->eval( bindings->cdr()->car());
	newEnv->put( key, value);
     }

     // evaluate expressions in new environment
     ctx.eval->pushEnvironment( newEnv);
     Handle_ptr pos, evaluated = 0;

#if defined( DEBUG) && DEBUG > 2
     std::cerr << "let expressions == ";
     printList( expr->cdr()->cdr(), std::cerr);
     std::cerr << std::endl;
#endif 

     for (pos=expr->cdr()->cdr(); !eq( pos, ctx.NIL); pos=pos->cdr())
        evaluated = ctx.eval->eval( pos->car());
     ctx.eval->popEnvironment();

     MCAssert( 0 != evaluated, "let failed to evaluate body");
     return evaluated;

  } else {
     throw ArgumentCountException( -2, __FILE__, __LINE__);
  }

  MCAssertNotReached( 0);
}


Handle_ptr
LetStarCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  MCAssertValidInstance();
  MCAssert( 0 != expr && !eq( expr, ctx.NIL), "illegal expression");

  if (!eq( expr->cdr(), ctx.NIL) && !eq( expr->cdr()->cdr(), ctx.NIL)) {
     // create new binding environment which will be used as current env
     ctx.eval->pushEnvironment();
     Handle_ptr bindList = expr->cdr()->car();

#if defined( DEBUG) && DEBUG > 2
     std::cerr << "let bindings == ";
     printList( bindList, std::cerr);
     std::cerr << std::endl;
#endif

     for ( ; !eq( bindList, ctx.NIL); bindList = bindList->cdr()) {
        Handle_ptr bindings = bindList->car();
#if defined( DEBUG) && DEBUG > 3
        std::cerr << " evaluate binding ";
	printList( bindings, std::cerr);
	std::cerr << std::endl;
#endif
	std::string key = bindings->car()->stringValue();
	Handle_ptr value = ctx.eval->eval( bindings->cdr()->car());
	ctx.eval->currentEnvironment()->put( key, value);
     }

     // evaluate expressions in new environment
     Handle_ptr pos, evaluated = 0;

#if defined( DEBUG) && DEBUG > 2
     std::cerr << "let expressions == ";
     printList( expr->cdr()->cdr(), std::cerr);
     std::cerr << std::endl;
#endif 

     for (pos=expr->cdr()->cdr(); !eq( pos, ctx.NIL); pos=pos->cdr())
        evaluated = ctx.eval->eval( pos->car());
     ctx.eval->popEnvironment();

     MCAssert( 0 != evaluated, "let failed to evaluate body");
     return evaluated;

  } else {
     throw ArgumentCountException( -2, __FILE__, __LINE__);
  }

  MCAssertNotReached( 0);
}


Handle_ptr
IfCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  Handle_ptr args = expr->cdr();
  if (3 != listlength( args))
    throw ArgumentCountException( 3, __FILE__, __LINE__);

  if (ctx.eval->eval( args->car()) != ctx.FALSE)
    return ctx.eval->eval( args->cdr()->car());
  else
    return ctx.eval->eval( args->cdr()->cdr()->car());
}


Handle_ptr
AndCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  Handle_ptr args = expr->cdr();
  int argc = listlength( args);
  if (0 == argc)
    //FIXME: needs *at least* one argument
    throw ArgumentCountException( 1, __FILE__, __LINE__);

  int i;
  Handle_ptr evaluated = 0;
  for (i=0; i<argc; ++i, args=args->cdr()) {
    evaluated = ctx.eval->eval( args->car());
    if (eq( evaluated, ctx.FALSE))
      return ctx.FALSE;
  }
  assert( 0 != evaluated);
  return evaluated;
}

Handle_ptr
OrCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
  Handle_ptr args = expr->cdr();
  int argc = listlength( args);
  if (0 == argc)
    //FIXME: needs *at least* one argument
    throw ArgumentCountException( 1, __FILE__, __LINE__);

  int i;
  Handle_ptr evaluated = 0;
  for (i=0; i<argc; ++i, args=args->cdr()) {
    evaluated = ctx.eval->eval( args->car());
    if (!eq( evaluated, ctx.FALSE)) {
      assert( 0 != evaluated);
      return evaluated;
    }
  }

  return ctx.FALSE;
}

Handle_ptr
TimeCommand::execute( Context &ctx, Environment *env, Handle_ptr expr)
{
   clock_t stopwatch = clock();
   Handle_ptr pos;
   for (pos=expr->cdr(); !eq( pos, ctx.NIL); pos=pos->cdr())
      ctx.eval->eval( pos->car());
   stopwatch = clock()-stopwatch;

   return ctx.factory->makeDouble( static_cast<double>( stopwatch) /
   				   static_cast<double>( CLOCKS_PER_SEC));
}   


/*
 * FUNCTIONS
 */
Handle_ptr displayCommand( CBuiltinAdapter *adapter, Context &ctx,
			   Environment *env, std::vector<Handle_ptr> args)
{
  std::vector<Handle_ptr>::iterator pos;
  for (pos = args.begin(); pos != args.end(); ++pos)
     std::cout << *(*pos) << std::endl;
  return ctx.TRUE; //FIXME: what should be returned to be standards compliant?
}


Handle_ptr carCommand( CBuiltinAdapter *adapter, Context &ctx,
		       Environment *env, std::vector<Handle_ptr> args)
{
  if (1 != args.size())
    throw ArgumentCountException( 1, __FILE__, __LINE__);
  return args[0]->car();
}

Handle_ptr cdrCommand( CBuiltinAdapter *adapter, Context &ctx,
		       Environment *env, std::vector<Handle_ptr> args)
{
  if (1 != args.size())
    throw ArgumentCountException( 1, __FILE__, __LINE__);
  return args[0]->cdr();
}

Handle_ptr memstatCommand( CBuiltinAdapter *adapter, Context &ctx,
			   Environment *env, std::vector<Handle_ptr> args)
{
   if (args.size() > 0) {
     std::cerr << "warning: " << adapter->name()
	  << " doesn't require any arguments." << std::endl;
   }
   Handle_ptr freeCount = ctx.factory->makeCons( ctx.factory->makeInteger( ctx.factory->freeCount()));
   return ctx.factory->makeCons( ctx.factory->makeInteger( ctx.factory->size()), freeCount);
}


Handle_ptr gcCommand( CBuiltinAdapter *adapter, Context &ctx,
		      Environment *env, std::vector<Handle_ptr> args)
{
  //FIXME:
  if (args.size() > 0) {
    std::cerr << "warning: " << adapter->name()
	 << " doesn't require any arguments." << std::endl;
  }
  ctx.interp->gc();
  return ctx.TRUE;
}


Handle_ptr beginCommand( CBuiltinAdapter *adapter, Context &ctx,
			 Environment *env, std::vector<Handle_ptr> args)
{
  if (0 == args.size())
    throw ArgumentCountException( 1, __FILE__, __LINE__);
  return args[args.size()-1];
}

  
Handle_ptr consCommand( CBuiltinAdapter *adapter, Context &ctx,
			Environment *env, std::vector<Handle_ptr> args)
{
  if (2 != args.size())
    throw ArgumentCountException( 2, __FILE__, __LINE__);
  return ctx.factory->makeCons( args[0], args[1]);
}

TypePredicate::TypePredicate( Handle::NodeType t)
{
  m_type = t;
}

TypePredicate::~TypePredicate()
{
}

Handle_ptr
TypePredicate::execute( Context &ctx, Environment *env, std::vector<Handle_ptr> args)
{
  if (1 != args.size())
    throw ArgumentCountException( 2, __FILE__, __LINE__);
  return (args[0]->hasType( m_type))? ctx.TRUE : ctx.FALSE;
}

Handle_ptr
notCommand( CBuiltinAdapter *adapter, Context &ctx, Environment *env,
	    std::vector<Handle_ptr> args)
{
  if (1 != args.size())
    throw ArgumentCountException( 1, __FILE__, __LINE__);
  if (eq( args[0], ctx.FALSE))
    return ctx.TRUE;
  return ctx.FALSE;
}

Handle_ptr
versionCommand( CBuiltinAdapter *adapter, Context &ctx, Environment *env,
	        std::vector<Handle_ptr> args)
{
   if (0 != args.size()) {
      std::cerr << "warning: ignoring argument(s)" << std::endl;
   }
   return ctx.factory->makeString( LISPEL_VERSION);
}

Handle_ptr 
exitCommand( CBuiltinAdapter *adapter, Context &ctx, Environment *env,
	     std::vector<Handle_ptr> args)
{
   if (0 != args.size()) {
      std::cerr << "warning: " << adapter->name()
	   << " doesn't require any arguments" << std::endl;
   }
   ctx.interp->exit();
   return ctx.TRUE;
}

Handle_ptr
loadCommand( CBuiltinAdapter *adapter, Context &ctx, Environment *env,
	     std::vector<Handle_ptr> args)
{
   if (1 != args.size())
      throw ArgumentCountException( 1, __FILE__, __LINE__);
   std::string filename = args[0]->stringValue();
   return (ctx.interp->source( filename))? ctx.TRUE : ctx.FALSE;
}


void addCoreCommands( Interpreter &interp)
{
  // special forms
  interp.addSpecial( "and", new AndCommand());
  interp.addSpecial( "cond", new CondCommand());
  interp.addSpecial( "define", new DefineCommand());
  interp.addSpecial( "if", new IfCommand());
  interp.addSpecial( "lambda", new LambdaCommand());
  interp.addSpecial( "let", new LetCommand());
  interp.addSpecial( "let*", new LetStarCommand());
  interp.addSpecial( "or", new OrCommand());
  interp.addSpecial( "quote", new QuoteCommand());
  interp.addSpecial( "time", new TimeCommand());

  // some basic functions
  interp.addBuiltin( "begin", beginCommand);
  interp.addBuiltin( "car", carCommand);
  interp.addBuiltin( "cdr", cdrCommand);
  interp.addBuiltin( "cons", consCommand);
  interp.addBuiltin( "display", displayCommand);
  interp.addBuiltin( "eq?", new BinaryPredicateAdapter( eq));
  interp.addBuiltin( "eqv?", new BinaryPredicateAdapter( eqv));
  interp.addBuiltin( "exit", exitCommand);
  interp.addBuiltin( "gc", gcCommand);
  interp.addBuiltin( "load", loadCommand);
  interp.addBuiltin( "memstat", memstatCommand);
  interp.addBuiltin( "not", notCommand);
  interp.addBuiltin( "version",  versionCommand);

  // type predicates
  interp.addBuiltin( "boolean?", new TypePredicate( Handle::ntBOOLEAN));
  interp.addBuiltin( "char?", new TypePredicate( Handle::ntCHARACTER));
  interp.addBuiltin( "double?", new TypePredicate( Handle::ntDOUBLE));
  interp.addBuiltin( "integer?", new TypePredicate( Handle::ntINTEGER));
  interp.addBuiltin( "pair?", new TypePredicate( Handle::ntCONS));
  interp.addBuiltin( "string?", new TypePredicate( Handle::ntSTRING));
  interp.addBuiltin( "symbol?", new TypePredicate( Handle::ntSYMBOL));
  interp.addBuiltin( "hashtable?", new TypePredicate( Handle::ntHASHTABLE));
  interp.addBuiltin( "keyword?", new TypePredicate( Handle::ntKEYWORD));
  interp.addBuiltin( "set?", new TypePredicate( Handle::ntSET));
}

